lang/py/avro/io.py

```diff
@@ -199,13 +201,745 @@ def _map_iterator(node: ValidationNode) -> ValidationNodeGeneratorType:
+        pass
+
+    @abstractmethod
+    def write_union(self, datum_writer: "DatumWriter", writers_schema: avro.schema.UnionSchema, datum: object) -> None:
+        pass
+
+    @abstractmethod
+    def write_record(self, datum_writer: "DatumWriter", writers_schema: avro.schema.RecordSchema, datum: Mapping[str, object]) -> None:
+        pass
+
+
+class Decoder(ABC):
+
+    @abstractmethod
+    def read_null(self) -> None:
+        pass
+
+    @abstractmethod
+    def read_boolean(self) -> bool:
+        pass
+
+    @abstractmethod
+    def read_int(self) -> int:
+        pass
+
+    @abstractmethod
+    def read_long(self) -> int:
+        pass
+
+    @abstractmethod
+    def read_float(self) -> float:
+        pass
+
+    @abstractmethod
+    def read_double(self) -> float:
+        pass
+
+    @abstractmethod
+    def read_decimal_from_bytes(self, precision: int, scale: int) -> decimal.Decimal:
+        pass
+
+    @abstractmethod
+    def read_decimal_from_fixed(self, precision: int, scale: int, size: int) -> decimal.Decimal:
+        pass
+
+    @abstractmethod
+    def read_bytes(self) -> bytes:
+        pass
+
+    @abstractmethod
+    def read_utf8(self) -> str:
+        pass
+
+    @abstractmethod
+    def read_date_from_int(self) -> datetime.date:
+        pass
+
+    @abstractmethod
+    def read_time_millis_from_int(self) -> datetime.time:
+        pass
+
+    @abstractmethod
+    def read_time_micros_from_long(self) -> datetime.time:
+        pass
+
+    @abstractmethod
+    def read_timestamp_millis_from_long(self) -> datetime.datetime:
+        pass
+
+    @abstractmethod
+    def read_timestamp_micros_from_long(self) -> datetime.datetime:
+        pass
+
+    @abstractmethod
+    def read_fixed(self, writers_schema: avro.schema.FixedSchema, readers_schema: avro.schema.Schema) -> bytes:
+        pass
+
+    @abstractmethod
+    def read_union(self, datum_reader: "DatumReader", writers_schema: avro.schema.UnionSchema, readers_schema: avro.schema.UnionSchema) -> object:
+        pass
+
+    @abstractmethod
+    def read_enum(self, datum_reader: "DatumReader", writers_schema: avro.schema.EnumSchema, readers_schema: avro.schema.EnumSchema) -> str:
+        pass
+
+    @abstractmethod
+    def read_array(self, datum_reader: "DatumReader", writers_schema: avro.schema.ArraySchema, readers_schema: avro.schema.ArraySchema) -> List[object]:
+        pass
+
+    @abstractmethod
+    def read_map(self, datum_reader: "DatumReader", writers_schema: avro.schema.MapSchema, readers_schema: avro.schema.MapSchema) -> Mapping[str, object]:
+        pass
+
+    @abstractmethod
+    def read_record(
+            self, datum_reader: "DatumReader", writers_schema: avro.schema.RecordSchema, readers_schema: avro.schema.RecordSchema) -> Mapping[str, object]:
+        pass
+
+
+class DatumReader:
+    """Deserialize Avro-encoded data into a Python data structure."""
+
+    _writers_schema: Optional[avro.schema.Schema]
```

Comment:
I know it's copied but I think it's wrong because it defines type for a class attribute but it's an object attribute. Probably should be removed.

There is the same pattern in multiple places in this file.

Reply:
This is totally fine:

https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#classes

Reply:
Looks really confusing to me. Especially that the example is wrong here:
```python
class MyClass:
    # This is an instance variable with a default value
    charge_percent: int = 100
```
It's still a class variable. It works here because the value is immutable. It won't work as expected if it will be something like `items: List[int] = [100]`.

I wouldn't support that kind of confusion.

Reply:
> It's still a class variable. It works here because the value is immutable. It won't work as expected if it will be something like items: List[int] = [100].

There is not assignment here, it is just the type declaration [PEP-0526](https://www.python.org/dev/peps/pep-0526/).

You're totally correct that a mutable value things could be broken though.